### PR TITLE
Loosen type constraints for deepEqual assertion

### LINKED ISSUE #2964: Support `Actual extends Expected` in conjunction with `Expected extends Actual`
### The Problem

The type `DeepEqualAssertion` requires that the `Expected` type extends the `Actual` type.
However, sometimes we want this relationship to be the other way around.

Consider the following example:

```ts
test("my test", (t) => {
  const expected = { foo: [1, 2, 3] } as const;
  const actual = functionToTest(1, 2, 3);

  t.deepEqual(actual, expected);
});
```

If the type of `actual` is inferred as `{ foo: number[] }` then this does not extend the type of `expected` which is `{ readonly foo: readonly number[] }`. Thus TypeScript reports an error.

### Suggested Fix

Add a overload to `DeepEqualAssertion` to handle this case:

```ts
export interface DeepEqualAssertion {
	/**
	 * Assert that `actual` is [deeply equal](https://github.com/concordancejs/concordance#comparison-details) to
	 * `expected`, returning a boolean indicating whether the assertion passed.
	 */
	<Actual, Expected extends Actual>(actual: Actual, expected: Expected, message?: string): actual is Expected;

	/**
	 * Assert that `actual` is [deeply equal](https://github.com/concordancejs/concordance#comparison-details) to
	 * `expected`, returning a boolean indicating whether the assertion passed.
	 */
	<Actual extends Expected, Expected>(actual: Actual, expected: Expected, message?: string): expected is Actual;

	/** Skip this assertion. */
	skip(actual: any, expected: any, message?: string): void;
}
```

### PR DESCRIPTION
**Pull request by RebeccaStevens**
fix #2964

### CODE REVIEW

**Comment by novemberborn** [comment]
Cheers @RebeccaStevens. Could you add a test for this under https://github.com/avajs/ava/tree/main/test-d?

File: types/assertions.d.ts [hunk 1]
```diff
@@ -137,6 +137,12 @@ export interface DeepEqualAssertion {
 	 * `expected`, returning a boolean indicating whether the assertion passed.
 	 */
 	<Actual, Expected extends Actual>(actual: Actual, expected: Expected, message?: string): actual is Expected;
+	
+	/**
+	 * Assert that `actual` is [deeply equal](https://github.com/concordancejs/concordance#comparison-details) to
+	 * `expected`, returning a boolean indicating whether the assertion passed.
+	 */
+	<Actual extends Expected, Expected>(actual: Actual, expected: Expected, message?: string): expected is Actual;
```
Comment (RebeccaStevens) [inline types/assertions.d.ts:145]: Actually this might not want to be a type guards. It might just simply want to return a boolean.

**Comment by RebeccaStevens** [comment]
> Cheers @RebeccaStevens. Could you add a test for this under [`main`/test-d](https://github.com/avajs/ava/tree/main/test-d)?

Sure thing. Give me a few minutes.

File: types/assertions.d.ts (hunk 1 above)
Comment (RebeccaStevens) [reply types/assertions.d.ts:145]: I'll figure out this when writing the tests.

**Comment by RebeccaStevens** [comment]
@novemberborn Done

File: test-d/deep-equal.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+});
+
+test('neither exectend each other', t => {
+	type Expected = {readonly foo: readonly [1, 2, 3]};
+	type Actual = {foo: number[]};
+	const expected: Expected = {foo: [1, 2, 3]};
+	const actual: Actual = {foo: [1, 2, 3]};
+	if (t.deepEqual(actual, expected)) {
+		t.pass();
+	}
```
Comment (novemberborn) [inline test-d/deep-equal.ts:31]: The `t.pass()` isn't necessary.

```suggestion
	t.deepEqual(actual, expected);
```

File: types/assertions.d.ts [hunk 3]
```diff
@@ -138,6 +138,18 @@ export interface DeepEqualAssertion {
...
+
+	/**
+	 * Assert that `actual` is [deeply equal](https://github.com/concordancejs/concordance#comparison-details) to
+	 * `expected`, returning a boolean indicating whether the assertion passed.
+	 */
+	<Actual, Expected>(actual: Actual, expected: Expected, message?: string): boolean;
```
Comment (novemberborn) [inline types/assertions.d.ts:151]: Why do we need this overload? Wouldn't the assertion always fail if neither extend one another?

Comment (RebeccaStevens) [reply types/assertions.d.ts:151]: If part of `Expected` extended part of `Actual`, but another part `Expected` is a super type of `Actual`, then neither extends the other but both my have the same value.

For example, imagine that one of the properties on `Expected` is typed as readonly, while the corresponding property on `Actual` isn't. And then for another property the two share, the readonlyness is declared the other way around.

**Comment by novemberborn** [comment]
Thanks @RebeccaStevens!

**Pull request merged**